File: examples/scala-sbt-example/src/test/scala/org/galaxio/performance/picatinny/feeders/FeederValidationFeeders.scala
package org.galaxio.performance.picatinny.feeders

import io.gatling.core.feeder.Feeder
import org.galaxio.gatling.feeders.faker.{Faker, GeneratedFeeder}
import org.galaxio.gatling.feeders.faker.Predef._

/** Feeder data for the feeder-validation e2e (test-model layer 4). Data only (galaxio-gatling-pro boundaries).
  *
  * `all` is a single ZIPPED picatinny feeder (`GeneratedFeeder` with every field) — one `.feed()` increments ALL values at
  * once, so the scenario pulls one record carrying every generated value rather than feeding each feeder separately. `patterns`
  * lists the EXPECTED value pattern per field; it drives the request headers, the mock echo body, and the response checks from
  * one source of truth.
  */
object FeederValidationFeeders {

  /** Faker `GeneratedFeeder` fields — values produced by `all`, one record per `.feed(all)`. */
  val fakerPatterns: List[(String, String)] = List(
    "uuid"       -> "[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}", // RFC-4122 v4, lowercase
    "str"        -> "[a-zA-Z0-9]{12}",
    "innPerson"  -> "\\d{12}",                                                             // natural person = 12 digits
    "innCompany" -> "\\d{10}",                                                             // legal entity = 10 digits
    "snils"      -> "\\d{11}",
    "pan"        -> "\\d{16}",
    "ogrn"       -> "\\d{13}",
    "ogrnip"     -> "\\d{15}",
    "kpp"        -> "\\d{9}",
    "passport"   -> "\\d{10}",
    "date"       -> "\\d{8}",
  )

  /** Non-deprecated picatinny-core feeders ([[org.galaxio.gatling.feeders]]) — values already in the session, fed by
    * `PicatinnyScenario`. Keys here are the ones that do NOT collide with `fakerPatterns` (uuid/pan/ogrn/kpp/snils/passport are
    * round-tripped under those same keys by the faker set); `RegexFeeder` is deprecated and the zipped/split feeders carry
    * composite keys, so all three are omitted. Loose-format phones assert non-blank round-trip only.
    */
  val picatinnyPatterns: List[(String, String)] = List(
    "currentDate"    -> "\\d{4}-\\d{2}-\\d{2}", // CurrentDateFeeder, ISO_LOCAL_DATE
    "randomDate"     -> "\\d{4}-\\d{2}-\\d{2}", // RandomDateFeeder, default yyyy-MM-dd
    "rangeFrom"      -> "\\d{4}-\\d{2}-\\d{2}", // RandomDateRangeFeeder lower bound
    "rangeTo"        -> "\\d{4}-\\d{2}-\\d{2}", // RandomDateRangeFeeder upper bound
    "digit"          -> "-?\\d+",               // RandomDigitFeeder, Int
    "customValue"    -> "custom-\\d+",          // CustomFeeder
    "phone"          -> ".+",                   // RandomPhoneFeeder, default format
    "e164Phone"      -> ".+",                   // RandomPhoneFeeder, E164
    "formattedPhone" -> ".+",                   // RandomPhoneFeeder, custom format
    "tollFreePhone"  -> ".+",                   // RandomPhoneFeeder, toll-free
    "randomString"   -> "[a-zA-Z0-9]{12}",      // RandomStringFeeder, alphanumeric(12)
    "rangeString"    -> "[abc123]{4,8}",        // RandomRangeStringFeeder
    "sequence"       -> "\\d+",                 // SequentialFeeder, Long
    "csvValue"       -> "alpha|beta|gamma",     // SeparatedValuesFeeder, circular
    "natItn"         -> "\\d{12}",              // RandomNatITNFeeder
    "jurItn"         -> "\\d{10}",              // RandomJurITNFeeder
    "psrnsp"         -> "\\d{15}",              // RandomPSRNSPFeeder
    "lambdaValue"    -> "[0-9a-f]{8}",          // feeder { } lambda, UUID prefix
    "greekLetter"    -> "alpha|beta|gamma",     // IndexedSeq.toFeeder
  )

  /** Every field round-tripped over HTTP: faker generators plus non-deprecated picatinny-core feeders. */
  val patterns: List[(String, String)] = fakerPatterns ++ picatinnyPatterns

  /** One zipped feeder: a single `.feed(all)` advances every generator together (one record, all fields). */
  val all: Feeder[Any] = GeneratedFeeder(
    "uuid"       -> Faker.uuid.string,
    "str"        -> Faker.string.alphanumeric(12),
    "innPerson"  -> Faker.ru.inn.person(),
    "innCompany" -> Faker.ru.inn.company(),
    "snils"      -> Faker.ru.snils(),
    "pan"        -> Faker.finance.pan(),
    "ogrn"       -> Faker.ru.ogrn(),
    "ogrnip"     -> Faker.ru.ogrnip(),
    "kpp"        -> Faker.ru.kpp(),
    "passport"   -> Faker.passport.ru(),
    "date"       -> Faker.date.today().format("yyyyMMdd"),
  )
}

File: examples/java-maven-example/src/test/java/org/galaxio/performance/picatinny/feeders/FeederValidationFeeders.java
package org.galaxio.performance.picatinny.feeders;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import static org.galaxio.gatling.javaapi.FakerApi.*;
import static org.galaxio.gatling.javaapi.Feeders.GeneratedFeeder;

public final class FeederValidationFeeders {

    private FeederValidationFeeders() {}

    /** Faker GeneratedFeeder fields — values produced by {@link #all()}. */
    public static final List<Map.Entry<String, String>> FAKER_PATTERNS = List.of(
        Map.entry("uuid",       "[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}"),
        Map.entry("str",        "[a-zA-Z0-9]{12}"),
        Map.entry("innPerson",  "\\d{12}"),
        Map.entry("innCompany", "\\d{10}"),
        Map.entry("snils",      "\\d{11}"),
        Map.entry("pan",        "\\d{16}"),
        Map.entry("ogrn",       "\\d{13}"),
        Map.entry("ogrnip",     "\\d{15}"),
        Map.entry("kpp",        "\\d{9}"),
        Map.entry("passport",   "\\d{10}"),
        Map.entry("date",       "\\d{8}")
    );

    /** Non-deprecated picatinny-core feeders — values already in the session, fed by PicatinnyScenario.
     *  Keys that collide with FAKER_PATTERNS (uuid/pan/ogrn/kpp/snils/passport) are round-tripped under those keys;
     *  RegexFeeder is deprecated and split/zipped feeders carry composite keys, so all are omitted. Loose-format
     *  phones assert non-blank round-trip only. */
    public static final List<Map.Entry<String, String>> PICATINNY_PATTERNS = List.of(
        Map.entry("currentDate",    "\\d{4}-\\d{2}-\\d{2}"),
        Map.entry("randomDate",     "\\d{4}-\\d{2}-\\d{2}"),
        Map.entry("rangeFrom",      "\\d{4}-\\d{2}-\\d{2}"),
        Map.entry("rangeTo",        "\\d{4}-\\d{2}-\\d{2}"),
        Map.entry("digit",          "-?\\d+"),
        Map.entry("customValue",    "custom-\\d+"),
        Map.entry("phone",          ".+"),
        Map.entry("e164Phone",      ".+"),
        Map.entry("formattedPhone", ".+"),
        Map.entry("tollFreePhone",  ".+"),
        Map.entry("randomString",   "[a-zA-Z0-9]{12}"),
        Map.entry("rangeString",    "[abc123]{4,8}"),
        Map.entry("sequence",       "\\d+"),
        Map.entry("csvValue",       "alpha|beta|gamma"),
        Map.entry("natItn",         "\\d{12}"),
        Map.entry("jurItn",         "\\d{10}"),
        Map.entry("psrnsp",         "\\d{15}")
    );

    /** Every field round-tripped over HTTP: faker generators plus non-deprecated picatinny-core feeders. */
    public static final List<Map.Entry<String, String>> PATTERNS = patterns();

    private static List<Map.Entry<String, String>> patterns() {
        List<Map.Entry<String, String>> all = new ArrayList<>(FAKER_PATTERNS);
        all.addAll(PICATINNY_PATTERNS);
        return List.copyOf(all);
    }

    /** Single zipped feeder — one `.feed(all())` advances every generator together. */
    public static Iterator<Map<String, Object>> all() {
        return GeneratedFeeder(
            field("uuid",       uuidString()),
            field("str",        alphanumeric(12)),
            field("innPerson",  innPerson()),
            field("innCompany", innCompany()),
            field("snils",      snils()),
            field("pan",        pan()),
            field("ogrn",       ogrn()),
            field("ogrnip",     ogrnip()),
            field("kpp",        kpp()),
            field("passport",   passportRu()),
            field("date",       formatDate(datePast(1), "yyyyMMdd"))
        );
    }
}
